### DISCUSSION TITLE
City Hub Broadcast and receive

### DISCUSSION THREAD

**Discussion by wallabyroad** [pybricks/support#1575] [General]
I have been trying for days to connect my send and receive color data via two CityHubs, To try and replicate the latest collision avoidance as demonstrated [here:](https://www.youtube.com/watch?v=n7kxz8-DZwA 
)
)Having used all the online examples provided to no avail. My programming knowledge is still rather quite basic and it is becoming increasingly frustrating.

My trains generally run on the following code but I would like all the trains to be able to talk to each other. Some assistance would be greatly appreciated.

![code example](https://github.com/pybricks/support/assets/404184/969473a8-93ab-4c5e-9ea4-6cdee3c3e622)

**Comment by laurensvalk** [comment]
Thanks for asking! Both trains in that video use a variation of this program:

![image](https://github.com/pybricks/support/assets/12326241/09eed7aa-2fe3-46a9-8a83-0cf1ebca0630)

The train in this program broadcasts on channel 34 and listens on channel 12. The program for the other train is exactly the same, but it broadcasts on channel 12 and listens on channel 34.

You can learn more about the techniques used in this program in these tutorials, which you might like to try out separately first:
- [How to make smart Lego Train layouts using color markers](https://www.youtube.com/watch?v=kwnSbnlhOkA)
- [Hub to hub communication](https://www.youtube.com/watch?v=WzmcihSV2YE)

**Reply by laurensvalk** [reply]
Here is the simpler version for a single train without Bluetooth. You might like to try this first:

![image](https://github.com/pybricks/support/assets/12326241/42711ce0-f230-456a-b447-c449ed3fd666)

**Reply by wallabyroad** [reply]
Thank you for the reply. I will try these out and let you know how it goes

**Comment by trustMeIAmANinja** [comment]
I am continuing the discussion about missed BLE broadcasts from Reddit (https://www.reddit.com/r/LEGOtrains/comments/1bq45lc/comment/kz931lx/)

Here is the sample code I created for checking message drops/delays

The sender side
=================================
![image](https://github.com/pybricks/support/assets/79006047/61e6aa95-d9a8-4ecf-b66f-76be74c7d53e)

The Receiver Side
=================================
![image](https://github.com/pybricks/support/assets/79006047/562854c7-bd9b-4032-ba54-cd25ad2cdf8c)


================================
Here is console output from a run.. https://gist.github.com/trustMeIAmANinja/67101a59689d9acd693a8493a30456c1
You can see every once is a while there are long delays of upto 10 seconds. I had a version where I was printing everything on the receiving side and during the delays, the receiver was getting None (as is nothing was being broadcast on the channel)

**Reply by laurensvalk** [reply]
Thank you for following up here! Does this also happen if you disconnect from the computer? Perhaps using a motor with `track_target(10 * recv_counter)` as the "display" since you can't see anything while disconnected.

**Reply by laurensvalk** [reply]
Is there a reason you kept the `broadcast(None)` plus wait?

I had that in my original example since I didn't need to broadcast for a few seconds. In your case though, everything seems to run in a fast loop, so you might as well skip the none command cut the number of operations to the bluetooth chip in half (to keep it less busy).

**Comment by trustMeIAmANinja** [comment]
I am going to put this as user error 🤦. I thought I had the Hubs disconnected from the laptop when i was struggling with the train demo. 
Following your suggestion above, I implemented a simple delay indicator using the hub light. Downloaded the code and ensured both hubs are disconnected from the laptop and started it off. they have running for about 30 mins now and all see is green. an occasional magenta, but overall nothing as bad as what i was seeing earlier.

![image](https://github.com/pybricks/support/assets/79006047/233e7280-f579-491a-ab12-836123e95ffa)

**Reply by laurensvalk** [reply]
Thanks so much for the update. Glad it is now working as expected. We're working on a site with some more ready-made examples so this stuff will be easier to find. Thanks for supporting Pybricks by using block coding :-)

**Reply by trustMeIAmANinja** [reply]
The two train collision avoidance is now working fine too, once I ensured both hubs were disconnected from Bluetooth. I love the block coding mode. I bought a personal license because I am going to be coaching a rookie FLL team this upcoming season and Pybricks makes it so much easier. The kids are learning Scratch and transitioning to FLL without Blocks coding would've been tricky.

**Reply by laurensvalk** [reply]
Awesome. If you have spare time with the team, doing something like you did with the trains is a great activity to learn all sorts of coding techniques without the immediate pressure of the competition. The color calibration technique might even come in handy for their robot too.
